src/Dialect/ONNX/ElementsAttrBuilder.hpp

```diff
@@ -0,0 +1,127 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===----------------------- ElementsAttrBuilder.hpp ----------------------===//
+//
+// ElementsAttrBuilder builds DisposableElementsAttr instances.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "src/Dialect/ONNX/DisposableElementsAttr.hpp"
+#include "src/Dialect/ONNX/DisposablePool.hpp"
+#include "src/Support/Strides.hpp"
+#include "src/Support/WideNum.hpp"
+
+#include <atomic>
+
+namespace onnx_mlir {
+
+// Builds DisposableElementsAttr instances.
+// Every instance is inserted into DisposablePool which garbage collects
+// unreachable instances between compiler passes.
+class ElementsAttrBuilder {
+public:
+  ElementsAttrBuilder(DisposablePool &disposablePool);
+
+  ElementsAttrBuilder(mlir::MLIRContext *context);
+
+  // Creates a DisposableElementsAttr instance backed by the data in membuf.
+  // The created instance takes ownership of membuf and will release it when the
+  // instance is disposed by garbage collection, unless it has shared membuf
+  // with other DisposableElementsAttr instances that are longer lived.
+  mlir::DisposableElementsAttr fromMemoryBuffer(
+      mlir::ShapedType type, std::unique_ptr<llvm::MemoryBuffer> membuf);
+
+  // Wraps elements in a DisposableElementsAttr if it isn't already a
+  // DisposableElementsAttr. If elements is DenseElementsAttr the wrapper
+  // points into elements' raw data, except if the element type is bool, then
+  // a deep copy is made that unpacks the bits because DisposableElementsAttr
+  // doesn't bit pack bools.
+  mlir::DisposableElementsAttr fromElementsAttr(mlir::ElementsAttr elements);
+
+  template <typename T>
+  using Filler = std::function<void(llvm::MutableArrayRef<T>)>;
+
+  // Constructs a DisposableElementsAttr and calls wideDataFiller to populate
+  // its memory buffer.
+  mlir::DisposableElementsAttr fromWideNums(
+      mlir::ShapedType type, const Filler<WideNum> &wideDataFiller);
+
+  // Constructs a DisposableElementsAttr and calls arrayFiller to populate
+  // its memory buffer.
+  template <typename T>
+  mlir::DisposableElementsAttr fromArray(
+      mlir::ShapedType type, const Filler<T> &arrayFiller);
+
+  // A transformer mutates elements.
+  using Transformer = std::function<void(llvm::MutableArrayRef<WideNum>)>;
+
+  // Constructs a transformer that changes every element to the result of
+  // applying the given function to the element.
+  template <typename UnaryFunction = std::function<WideNum(WideNum)>>
+  static Transformer functionTransformer(UnaryFunction fun);
+
+  // Returns a DisposableElementsAttr where each element is transformed
+  // by running the given transformer on all the elements.
+  mlir::DisposableElementsAttr transform(mlir::DisposableElementsAttr elms,
+      mlir::Type transformedElementType, Transformer transformer);
+
+  // Returns a DisposableElementsAttr that is the result of applying
```

Comment:
Can you specify which calls below reuse the original copy of the data and only add an "accessor" method that will compute the actual values when executed? probably all, so you could just state that.